README.md

studentlife: Tidy Handling and Navigation of a Valuable Mobile-Health Dataset
======================================================================================================

<img src="man/figures/logo.png" align="right" alt="" width="120" />

<!-- Badges Start -->
[![CRAN\_Release\_Badge](http://www.r-pkg.org/badges/version-ago/studentlife)](https://CRAN.R-project.org/package=studentlife)
[![devel Version](https://img.shields.io/badge/devel-1.1.0-brightgreen.svg)](https://github.com/frycast/studentlife)
[![Downloads](http://cranlogs.r-pkg.org/badges/grand-total/studentlife)](https://CRAN.R-project.org/package=studentlife)
[![Travis-CI Build Status](https://travis-ci.org/frycast/studentlife.svg?branch=master)](https://travis-ci.org/frycast/studentlife) 
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/frycast/studentlife/master?urlpath=rstudio) 
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0) 
[![Code coverage](https://codecov.io/gh/frycast/studentlife/branch/master/graph/badge.svg)](https://codecov.io/github/frycast/studentlife?branch=master)
[![Lifecycle Status](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3371922.svg)](https://doi.org/10.5281/zenodo.3371922)
<!-- Badges End -->

This package is [available on CRAN](https://cran.r-project.org/web/packages/studentlife/index.html).

Use this R package to download, navigate and analyse the Student-Life dataset. The Student-Life dataset contains 
    passive and automatic sensing data from the phones of a class 
    of 48 de-identified Dartmouth college students. It was collected 
    over a 10 week term. Additionally, the dataset contains Ecological 
    Momentary Assessment results along with pre- and post-study mental 
    health surveys, such as the PHQ-9. The intended use is to allow
    researchers and other interested people to assess 
    mental health, academic performance and behavioral trends. 
    The raw dataset and additional information is available at <https://studentlife.cs.dartmouth.edu/>.




Installation
------------

``` r
## Uncomment to install the package from CRAN
# install.packages("studentlife")

## Or, uncomment to install the package from GitHub
# install.packages("devtools")
# devtools::install_github("frycast/studentlife")
```

This studentlife repository includes a small sample dataset that can be used for practice and testing. You can download the sample data within R:
```
library(studentlife)
d <- tempdir()
download_studentlife(location = d, url = "testdata")
```

This README will use the studentlife data in its original format. 
Details on the full original dataset are available [here](https://studentlife.cs.dartmouth.edu).
However, an RData version of the dataset has been hosted on Zenodo as of 09/11/2019 [here](https://zenodo.org/record/3529253).
To download and use this RData version (which is faster to read and takes up less space):

```r
d <- tempdir()
download_studentlife(location = d, url = "rdata")

# Choose the schema and table from the list SL_tables:
SL_tables

# Example with activity table from sensing schema
schema <- "sensing"
table <- "activity"
act <- readRDS(paste0(d, "/dataset_rds/", schema, "/", table, ".Rds"))
act
```

In this README we will use the full dataset rather than just the sample data. The download size is 5 GB.

``` r
download_studentlife(location = d)
```

Use the interactive menu to browse the tables and schemas of the downloaded dataset:

``` r
tab <- studentlife::load_SL_tibble(location = d)
```

The object returned by the above function is called a
'StudentLife tibble' (or `SL_tbl`).
Restrictions can be placed on the menu options with `time_options`:

``` r
tab_t <- load_SL_tibble(location = d, time_options = "timestamp", csv_nrows = 10)
tab_p <- load_SL_tibble(location = d, time_options = "interval" , csv_nrows = 10)
tab_d <- load_SL_tibble(location = d, time_options = "dateonly" , csv_nrows = 10)
tab_s <- load_SL_tibble(location = d, time_options = "dateless" , csv_nrows = 10)
```

The `regularise_time` function can be used to summarise information within blocks of time,
producing an object called a 'regularised StudentLife tibble' (or `reg_SL_tbl`):

``` r
tab <- load_SL_tibble(
  loc = d, schema = "sensing", table = "activity", csv_nrows = 10)
  
regularise_time(
  tab, blocks = c("day","weekday"),
  act_inf = max(activity_inference), add_NAs = FALSE)
```

Produce a histogram showing PAM EMA response frequencies over the course of the study:

``` r
tab_PAM <- load_SL_tibble(schema = "EMA", table = "PAM", location = d)
response_hour_hist(tab_PAM, break_hours = 10)
```

![](man/figures/response_hour_histogram.png)

A summary will produce details such as EMA question, UIDs of 
dropped students, schema name, table name, and summary statistics.

```r
summary(tab_PAM)
```

``` 
EMA_questions 
 Refer to: Pollak, J. P., Adams, P., & Gay, G. (2011, May). 
 PAM: a photographic affect meter for frequent, in situ measurement of affect. 
 In Proceedings of the SIGCHI conference on Human factors in computing systems 
 (pp. 725-734). ACM. 

dropped_students 
 None 

column_names 
 picture_idx timestamp uid 

schema 
 EMA 

table 
 PAM 

Skim summary statistics
 n obs: 9040 
 n variables: 3 

-- Variable type:factor --------------------------------------------------------
 variable missing complete    n n_unique                        top_counts ordered
      uid       0     9040 9040       49 59: 437, 0: 390, 19: 384, 57: 377   FALSE

-- Variable type:numeric -------------------------------------------------------
    variable missing complete    n    mean         sd      p0     p25     p50      p75     p100
 picture_idx       0     9040 9040 8.85          4.17 1       6       8       12       16      
   timestamp       0     9040 9040 1.4e+09 1608614.45 1.4e+09 1.4e+09 1.4e+09  1.4e+09  1.4e+09
     hist
 ▂▃▆▇▃▆▅▅
 ▇▇▆▅▃▁▁▁
```

After regularising a StudentLife tibble, we can visualise the 
missing values in each block for each student:

```r
reg_PAM <- regularise_time(tab_PAM, blocks = c("day", "epoch"), m = mean(picture_idx, na.rm = TRUE))
vis_NAs(reg_PAM, response = "m")
```
![](man/figures/vis_NAs.png)

We can also visualise and compare the total number of responses 
received from each student over the course of the study:

```r
vis_response_counts(reg_PAM, response = "m")
```

![](man/figures/response_counts.png)


Summaries of survey data are formatted with question information and some answer statistics:

```r 
tab_PHQ9 <- load_SL_tibble(loc = d, schema = "survey", table = "PHQ-9")
summary(tab_PHQ9)
```

```
time_info 
 none 

survey_questions 
 Q1: Little interest or pleasure in doing things
 Q2: Feeling down, depressed, hopeless.
 Q3: Trouble falling or staying asleep, or sleeping too much.
 Q4: Feeling tired or having little energy
 Q5: Poor appetite or overeating
 Q6: Feeling bad about yourself or that you are a failure or have let yourself or your family down
 Q7: Trouble concentrating on things, such as reading the newspaper or watching television
 Q8: Moving or speaking so slowly that other people could have noticed. Or the opposite being so figety or restless that you have been moving around a lot more than usual
 Q9: Thoughts that you would be better off dead, or of hurting yourself
 Q10: Response
 

column_names 
 uid type Q1 Q2 Q3 Q4 Q5 Q6 Q7 Q8 Q9 Q10 

schema 
 survey 

table 
 PHQ-9 

skim 
Skim summary statistics
 n obs: 84 
 n variables: 12 

-- Variable type:factor --------------------------------------------------------
 variable missing complete  n n_unique                        top_counts ordered
       Q1       0       84 84        4 Not: 42, Sev: 28, Mor: 10, Nea: 4   FALSE
      Q10       5       79 84        4   Som: 37, Not: 36, NA: 5, Ext: 3   FALSE
       Q2       0       84 84        4  Not: 40, Sev: 34, Mor: 6, Nea: 4   FALSE
       Q3       0       84 84        4 Not: 41, Sev: 25, Mor: 10, Nea: 8   FALSE
       Q4       0       84 84        4 Sev: 38, Not: 23, Mor: 17, Nea: 6   FALSE
       Q5       0       84 84        4  Sev: 35, Not: 34, Mor: 9, Nea: 6   FALSE
       Q6       0       84 84        4  Not: 45, Sev: 27, Mor: 6, Nea: 6   FALSE
       Q7       0       84 84        4  Not: 48, Sev: 26, Mor: 7, Nea: 3   FALSE
       Q8       0       84 84        4  Not: 64, Sev: 15, Mor: 4, Nea: 1   FALSE
       Q9       0       84 84        3    Not: 75, Sev: 6, Mor: 3, NA: 0   FALSE
     type       0       84 84        2           pre: 46, pos: 38, NA: 0   FALSE
      uid       0       84 84       46            0: 2, 1: 2, 2: 2, 3: 2   FALSE
```


Software Testing
----------------

This studentlife repository includes many automated software tests implemented via [`testthat`](https://cran.r-project.org/web/packages/testthat/index.html). We use these to check for bugs before releasing new updates. They can be found under the directory `tests`. Also, we use [Travis-CI](https://travis-ci.org/frycast/studentlife) for continuous integration.

Community Guidelines
----------------
* Please give your feedback and report bugs at the [issues page](https://github.com/frycast/studentlife/issues).
* Contributions are welcome! The best way to contribute is to fork the project on GitHub, make a contribution to your fork, and then submit a pull request. A useful guide can be [found here](https://akrabat.com/the-beginners-guide-to-contributing-to-a-github-project/#summary).
* If you have questions or need support please email Daniel Fryer via the email on his [GitHub profile](https://github.com/frycast/).


<!--
DOCUMENTATION CHECKLIST
    A statement of need: Do the authors clearly state what problems the software is designed to solve and who the target audience is?
    Installation instructions: Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
    Example usage: Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
    Functionality documentation: Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
    Automated tests: Are there automated tests or manual steps described so that the function of the software can be verified?
    Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support
A statement of need
The authors should clearly state what problems the software is designed to solve and who the target audience is.
Installation instructions
There should be a clearly-stated list of dependencies. Ideally these should be handled with an automated package management solution.
    Good: A package management file such as a Gemfile or package.json or equivalent
    OK: A list of dependencies to install
    Bad (not acceptable): Reliance on other software not listed by the authors
Example usage
The authors should include examples of how to use the software (ideally to solve real-world analysis problems).
API documentation
Reviewers should check that the software API is documented to a suitable level.
    Good: All functions/methods are documented including example inputs and outputs
    OK: Core API functionality is documented
    Bad (not acceptable): API is undocumented
    
Community guidelines
There should be clear guidelines for third-parties wishing to:
    Contribute to the software
    Report issues or problems with the software
    Seek support
Functionality
Reviewers are expected to install the software they are reviewing and to verify the core functionality of the software.
Tests
Authors are strongly encouraged to include an automated test suite covering the core functionality of their software.
    Good: An automated test suite hooked up to an external service such as Travis-CI or similar
    OK: Documented manual steps that can be followed to objectively check the expected functionality of the software (e.g. a sample input file to assert behaviour)
    Bad (not acceptable): No way for you the reviewer to objectively assess whether the software works
-->





NEWS.md
# studentlife 1.1.0
Bug fixes and updates:
  
  * fixes timezone bug so that dates and times are correct
  * introduces PAM_quadrant function
  * load_SL_tibble will automatically call PAM_quadrant on PAM picture_idx
  * block name "hour" changes to "hour-in-day" to avoid conflict
  * vis_response_counts returns named counts vector
  * vis_response_counts plots in decreasing order
  * Adds RData format download link to download_studentlife
  * Updates documentation to explain RData download and usage from Zenodo
  * Updates dependency to skimr >= 1.0.7
  * Corrects regularise_time output for "day" and "hour"
  * Makes sure reg_SL_tbl class is returned by regularise_time
  * Add blocks to default title of vis_NAs plot
  * SL_tbl_load now guesses schema from table
  * removes pesky class-dropping tamper generics
  * provide option to remove safety checks in tibble transforms
  
# studentlife 1.0.3
This release only updates the authorship to give
correct authorship information.

  * Updated authorship information in DESCRIPTION

# studentlife 1.0.1
This release includes changes recommended by 
reviewers at the Journal of Open Source Software.

  * Fixed typos in documentation.
  * Include community guidelines in the README.md

# studentlife 1.0.0 (current version on CRAN)
This is the first release.


R/SL_tbl_load.R
#' download_studentlife
#'
#' Download the entire StudentLife dataset or
#' a smaller sample dataset for testing.
#'
#' If \code{url = "rdata"} then data will be downloaded
#' from <https://zenodo.org/record/3529253>
#' If \code{url = "dartmouth"} then data will be downloaded
#' from <https://studentlife.cs.dartmouth.edu/dataset/dataset.tar.bz2>
#' If \code{url = "testdata"} then data will be downloaded
#' from the test data at the studentlife GitHub repository
#' <https://github.com/frycast/studentlife>
#'
#'
#'@param url A character string. Either
#'"rdata" for the URL to the (more efficient)
#'RData format version hosted on Zenodo, or
#'"dartmouth" for the (original) Dartmouth URL, or
#'"testdata" for a small sample dataset. Otherwise
#'a full URL of your choice can be specified leading to
#'the StudentLife dataset as a \code{.tar.gz} file.
#'@param location The destination path. If the path does
#'not exist it is created with \code{\link{dir.create}}
#'@param unzip Logical. If \code{TRUE} then the
#'dataset will be unzipped with \code{\link[R.utils]{bunzip2}}.
#'Leave as default unless you plan to do it manually.
#'@param untar Logical. If \code{TRUE} then the
#'dataset will be untarred with \code{\link[utils]{untar}}.
#'Leave as default unless you plan to do it manually.
#'
#'@examples
#'d <- tempdir()
#'download_studentlife(location = d, url = "testdata")
#'
#'\donttest{
#'## With menu
#'load_SL_tibble(location = d)
#'}
#'
#'## Without menu
#'SL_tables
#'load_SL_tibble(schema = "EMA", table = "PAM", location = d)
#'
#'@export
download_studentlife <- function(
  url = "dartmouth",
  location = ".",
  unzip = TRUE,
  untar = TRUE) {

  zip <- FALSE
  if (tolower(url) == "dartmouth") {

    url <- paste0("https://studentlife.cs.dartmouth.edu",
                  "/dataset/dataset.tar.bz2")

    mes1 <- "Downloading the original StudentLife dataset..."
    f <- "dataset.tar.bz2"

  } else if (tolower(url) == "testdata") {

    url <- paste0("https://raw.githubusercontent.com/",
                  "frycast/studentlife/master/tests/",
                  "testthat/testdata/sample/sample_dataset.tar.bz2")

    mes1 <- "Downloading the small sample dataset..."
    f <- "dataset.tar.bz2"

  } else if (tolower(url) == "rdata") {

    url <- paste0("https://zenodo.org/record/3529253/",
                  "files/dataset_rds.zip?download=1")

    mes1 <- "Downloading the RData studentlife dataset..."
    f <- "dataset_rds.zip"
    untar <- FALSE
    zip <- TRUE

  } else {

    mes1 <- "Downloading from user specified url..."
    f <- "dataset.tar.bz2"

  }

  message(mes1)
  p <- paste0(location, "/", f)
  if (!dir.exists(location)) dir.create(location)
  if (!zip) {utils::download.file(url = url, destfile = p, cacheOK = FALSE)}
  if (zip) {utils::download.file(url = url, destfile = p, method = 'curl')}
  message("Download complete")

  if (zip && unzip) {
    message("Unzipping the dataset...")
    utils::unzip(p, exdir = location)
    message("Unzip complete")
  } else if (unzip) {
    message("Unzipping the dataset...")
    R.utils::bunzip2(p, remove = FALSE, skip = TRUE)
    message("Unzip complete")
  }
  if (untar) {
    f <- "dataset.tar"
    p <- paste0(location, "/", f)
    message("Untarring the dataset...")
    utils::untar(p, exdir = location)
    message("Untar complete")
  }
}


#' load_SL_tibble
#'
#' Import a chosen StudentLife table as
#' a tibble. Leave \code{schema} and \code{table}
#' unspecified to choose interactively via a
#' menu. This function is only intended for use
#' with the studentlife dataset in it's original
#' format, with the original directory structure.
#' See the examples below for the recommended alternative approach
#' to loading tables when the RData format is used.
#'
#' @param schema A character string. The menu 1 choice. Leave
#' blank to choose interactively.
#' @param table A character string. The menu 2 choice. Leave
#' blank to choose interactively.
#' @param location The path to a copy of the StudentLife dataset.
#' @param time_options A character vector specifying which
#' table types (out of "interval", "timestamp", "dateonly" and "dateless")
#' to include in the menu. This allows you to restrict menu options
#' according to the amount of date-time information present in the data.
#' The default includes all data. Note
#' this parameter only has an effect when used with the interactive menu.
#' @param vars Character vector of variable
#' names to import for all students. Leave
#' blank and this will be chosen interactively
#' if necesssary. If \code{vars} contains
#' "timestamp" then effort will be made
#' to convert "timestamp" to appropriate
#' variable name(s) for the target table.
#' @param csv_nrows An integer specifying the number of rows to read
#' per student if the target is a csv. The largest files in StudentLife are csv
#' files, so this allows code testing with less overhead.
#' @param datafolder Specifies the subfolder of \code{location}
#' that contains the relevant data. This should normally
#' be left as the default.
#' @param uid_range An integer vector. The range of uids in
#' the StudentLife study.
#'
#' @return
#' An object of class \code{SL_tibble} is returned. These inherit
#' properties from class \code{\link[tibble]{tibble}} and
#' class \code{\link{data.frame}}.
#' Depending on the date-time information available, the object
#' may also be a \code{timestamp_SL_tibble},
#' \code{interval_SL_tibble} or
#' \code{dateonly_SL_tibble} (which are all
#' subclasses of \code{SL_tibble}).
#'
#'@examples
#'## Example that uses RData format to efficiently
#'## download and load tables, as an alternative
#'## to using this function.
#'\donttest{
#' d <- tempdir()
#' download_studentlife(location = d, url = "rdata")
#'
#' # Choose the schema and table from the list SL_tables:
#' SL_tables
#'
#' # Example with activity table from sensing schema
#' schema <- "sensing"
#' table <- "activity"
#' act <- readRDS(paste0(d, "/dataset_rds/", schema, "/", table, ".Rds"))
#' act
#'}
#'
#'## Example that uses the studentlife dataset in
#'## its original format.
#'
#'# Use url = "dartmouth" for the full original dataset
#'d <- tempdir()
#'download_studentlife(location = d, url = "testdata")
#'
#'\donttest{
#'## With menu
#'load_SL_tibble(location = d)
#'}
#'
#'## Without menu
#'SL_tables
#'PAM <- load_SL_tibble(schema = "EMA", table = "PAM", location = d)
#'
#'## Load less data for testing with less overhead
#'act <- load_SL_tibble(schema = "sensing", table = "activity",
#'                      location = d, csv_nrows = 10)
#'
#'\donttest{
#'## Browse all tables with timestamps (non-interval)
#'load_SL_tibble(location = d, time_options = "timestamp")
#'
#'## Browse all tables with intervals
#'load_SL_tibble(location = d, time_options = "interval")
#'
#'## Browse all dateless tables
#'load_SL_tibble(location = d, time_options = "dateless")
#'}
#' @export
load_SL_tibble <- function(
  schema, table, location = ".",
  time_options = c("interval", "timestamp", "dateonly", "dateless"),
  vars, csv_nrows, datafolder = "dataset",
  uid_range = getOption("SL_uids")) {

  time_options <- tolower(time_options)
  datafolder <- paste0("/", datafolder)

  opt <- c("interval", "timestamp", "dateonly", "dateless")
  options_check(par = time_options, opt = opt)

  if ( !dir.exists(location) )
    stop("the directory specified by location parameter does not exist")

  if ( !dir.exists(paste0(location, datafolder)) )
    stop(paste0("the location does not have a subfolder named ", datafolder))

  if (!missing(vars)) {
    if( !("uid" %in% vars ) ) vars <- c("uid", vars)
  }

  if ( missing(schema) & !missing(table) ) {

    guess_schema <- names(
      unlist(lapply(studentlife::SL_tables, function(x){which(x == table)})))
    if (length(guess_schema) != 1) {
      stop(paste0("Had trouble determining the schema, ",
                  "try specifying it with the schema argument"))
    } else {
      schema <- guess_schema
    }

  }

  if (!missing(schema))
    schema <- pmatch(tolower(schema), tolower(menu_data$menu1_choices))

  if (!missing(table))
    table <- pmatch(tolower(table), tolower(menu_data$menu2_list[[schema]]))

  location <- paste0(location, datafolder)

  path <- get_path(location, schema, table, time_options)

  if ( path %in% menu_data$EMA_json ) {

    tab <- get_EMA_tab(path, location, vars)

  } else if ( path %in% menu_data$long_csv ) {

    tab <- get_long_csv_tab(path, location, vars, csv_nrows)

  } else if ( path %in% menu_data$wide_csv ) {

    tab <- get_wide_csv_tab(path, location, vars, csv_nrows)

  } else if ( path %in% menu_data$txt ) {

    tab <- get_txt_tab(path, location, vars)

  }

  tab$uid <- factor(tab$uid, levels = uid_range)

  tab <- structure(
    tab, schema = attr(path, "schema"),
    table = attr(path, "table"))

  if ( attr(tab, "table") == "PAM" && ("picture_idx" %in% names(tab)) ) {
    tab <- PAM_categorise(tab, pam_name = "picture_idx")
  }

  names(tab) <- clean_strings(names(tab))

  return(tab)
}


############################################################################
############################################################################
# Helper functions ---------------------------------

get_path <- function(location, menu1, menu2, time_options) {


  # Present interactive menu 1
  menu1_restrict <- unlist(
    menu_data$time_opt_list1[time_options], use.names = FALSE)
  menu1_choices <- menu_data$menu1_choices[
    which(menu_data$menu1_choices %in% menu1_restrict)]
  if ( missing(menu1) ) {

    menu1 <- menu1_choices[[utils::menu(
      choices = menu1_choices,
      title = "Choose Menu 1 option:")]]
  } else {

    menu1 <- menu1_choices[[menu1]]
  }

  schema <- menu1

  if (menu1 == "EMA") menu1 <- "EMA/response"

  # Present interactive menu 2
  menu2_choices <- menu_data$menu2_list[[menu1]]
  menu2_restrict <- unlist(
    menu_data$time_opt_list2[time_options], use.names = FALSE)
  menu2_choices <- menu2_choices[
    which(menu2_choices %in% menu2_restrict)]
  if ( missing(menu2) ) {

    menu2 <- utils::menu(
      choices = menu2_choices,
      title = "Choose Menu 2 option:")
  }

  if (is.null(menu2_choices))
    stop("No tables found in specified schema")

  menu2 <- menu2_choices[[menu2]]
  table <- menu2

  result <- paste0(menu1, "/", menu2)
  result <- get_name_from_path(result, split = 'other/')

  if ( result == "dining" ) result <- "dinning"

  result <- structure(result, schema = schema, table = table)

  return(result)
}

get_txt_tab <- function(path, location, vars) {

  `%>%` <- dplyr::`%>%`

  if ( !missing(vars) )
    if ( "timestamp" %in% vars )
      vars[pmatch("timestamp", vars)] <- "date-time"


  pr <- paste0(location, "/", path, "/", "u")
  paths <- c(paste0(pr, "0", seq(0,9), ".txt"),
             paste0(pr, seq(10,59), ".txt"))

  readr::read_csv(paths[2],
                  col_names = c("date-time","location","type"))
  tab <- list()
  missing_tab <- 0
  for (i in 1:60) {
    if(file.exists(paths[i])) {
      this_stud <- suppressMessages(
        readr::read_csv(paths[i], progress = FALSE,
                        col_names = c("date-time","location","type")))
      this_stud$uid <- i - 1
      tab[[length(tab)+1]] <- this_stud
    } else {
      missing_tab <<- missing_tab + 1
    }
  }

  # Bind students
  tab <- tab %>%
    dplyr::bind_rows() %>%
    tibble::as_tibble()

  if ( !missing(vars) ) tab <- dplyr::select(tab, vars)

  class(tab) <- c("SL_tbl", class(tab))

  if ( "date-time" %in% names(tab) ) {
    names(tab)[pmatch("date-time", names(tab))] <- "timestamp"
    tab$timestamp <- as.numeric(
      as.POSIXct(tab$timestamp, origin="1970-01-01",
                 tz = getOption("SL_timezone")))
    class(tab) <- c("timestamp_SL_tbl", class(tab))
  }

  return(tab)
}


get_wide_csv_tab <- function(path, location, vars, csv_nrows) {

  full_path <- paste0(location, "/", path, ".csv")
  name <- get_name_from_path(path)

  args <- list(file = full_path, na.strings = c("NA",""), check.names = FALSE)
  cn <- c("uid", "class1", "class2", "class3", "class4")
  if (name == "class") {args$col.names <- cn; args$header <- FALSE}
  if (name == "deadlines") {args$check.names <- TRUE}
  tab <- do.call(utils::read.csv, args)

  if ( name %in% menu_data$survey ) {
    exc <- which(names(tab) %in% c("uid","type"))
    q_text <- names(tab[,-exc])
    names(tab) <- c(names(tab[,exc]), paste0("Q", 1:length(names(tab[,-exc]))))
    attr(tab, "survey_questions") <- q_text
  }

  tab$uid <- as.integer(substr(tab$uid, 2, 3))
  tab <- tibble::as_tibble(tab)
  if(!missing(vars)) tab <- dplyr::select(tab, vars)


  if ( name %in% menu_data$dateonly ) {

    tab <- tidyr::gather(tab, "date", "deadlines", -c("uid"))
    tab$date <- as.Date(
      as.POSIXct(substr(tab$date,2,11), format = "%Y.%m.%d",
                 tz = getOption("SL_timezone")),
      tz = getOption("SL_timezone"))

    class(tab) <- c("SL_tbl", class(tab))

    class(tab) <- c("dateonly_SL_tbl", class(tab))

  } else {

    class(tab) <- c("SL_tbl", class(tab))

    class(tab) <- c("dateless_SL_tbl", class(tab))

  }

  return(tab)
}



get_long_csv_tab <- function(path, location, vars, csv_nrows) {

  `%>%` <- dplyr::`%>%`

  name <- get_name_from_path(path)
  if( name == "app_usage" ) name <- "running_app"
  if( name == "bluetooth" ) name <- "bt"
  paths <- generate_paths(location, path, name)

  if( !missing(vars) )
    if( "timestamp" %in% vars ) {
      if( name %in% menu_data$interval ) {
        if ( name == "conversation" ) {
          vars[pmatch("timestamp", vars)] <- "start_timestamp"
          vars <- c(vars, "end_timestamp")
        } else {
          vars[pmatch("timestamp", vars)] <- "start"
          vars <- c(vars, "end")
        }
      } else if ( name %in% c("bt","gps","wifi","wifi_location") ) {
        vars[pmatch("timestamp", vars)] <- "time"
      }
    }

  args <- list()
  if ( !missing(csv_nrows) ) args$nrows <- csv_nrows

  tab <- list()
  missing_tab <- 0
  if ( name == "wifi_location" || name == "gps" ) {
    args2 <- c(args, list(skip = 1, header = FALSE,
      stringsAsFactors = FALSE))
    for (i in 1:60) {
      if(file.exists(paths[i])) {
        args2$file <- paths[i]
        col_names <- as.character(
          utils::read.csv(file = paths[i], nrows = 1,
                          stringsAsFactors = FALSE,
                          header = FALSE))
        args2$col.names <- c(col_names, "to_drop")
        this_stud <- suppressMessages(
          do.call(utils::read.csv, args2))
        this_stud$to_drop <- NULL
        this_stud$uid <- i - 1
        tab[[length(tab)+1]] <- this_stud
      } else {
        missing_tab <<- missing_tab + 1
      }
    }
  } else {
    if ( missing(csv_nrows) ) csv_nrows <- Inf
    for (i in 1:60) {
      if(file.exists(paths[i])) {
        this_stud <- suppressMessages(
          readr::read_csv(
            file = paths[i], progress = FALSE, n_max = csv_nrows))
        this_stud$uid <- i - 1
        tab[[length(tab)+1]] <- this_stud
      } else {
        missing_tab <<- missing_tab + 1
      }
    }
  }

  if ( path == "sms" && missing(vars) ) {
    tab <- lapply(tab, function(x){
      x <- dplyr::select(
        as.data.frame(x), "id", "device", "timestamp", "uid")
    })
  }

  # Bind students
  tab <- tab %>%
    dplyr::bind_rows() %>%
    tibble::as_tibble()

  if ( !missing(vars) ) tab <- dplyr::select(tab, vars)

  if( name %in% menu_data$interval ) {
    if ( !(name == "conversation") ) {
      if ( "start" %in% names(tab) )
        names(tab)[pmatch("start", names(tab))] <- "start_timestamp"
      if ( "end" %in% names(tab) )
        names(tab)[pmatch("end", names(tab))] <- "end_timestamp"
    }
  } else if ( name %in% c("bt","gps","wifi","wifi_location") ) {
    if ( "time" %in% names(tab) )
      names(tab)[pmatch("time", names(tab))] <- "timestamp"
  }

  class(tab) <- c("SL_tbl", class(tab))

  if ("timestamp" %in% names(tab))
    class(tab) <- c("timestamp_SL_tbl", class(tab))

  if ("start_timestamp" %in% names(tab) && "end_timestamp" %in% names(tab))
    class(tab) <- c("interval_SL_tbl", class(tab))

  return(tab)
}



get_EMA_tab <- function(path, location, vars) {

  if (!missing(vars) ) {
    vars[pmatch("timestamp", vars)] <- "resp_time"
  }

  tab <- EMA_to_list(location, path)

  if ( missing(vars) ) {
    # Create a graphical menu to choose vars

    vars_list <- attributes(tab)$vars_present

    vars_opt <- unlist(lapply(vars_list, function(x) {
      vec <- unlist(x)
      vec[pmatch("resp_time", vec)] <- "timestamp"
      sort(paste0(vec, collapse = ", "))
    }))

    vars_opt <- vars_opt[order(nchar(vars_opt), decreasing = TRUE)]

    if ( length(vars_opt) > 1 ) {

      t <- paste0("Choose variables. ",
                  "Only students who share all chosen variables ",
                  "will be loaded. Choosing more variables usually ",
                  "implies more students will be discarded.")

      choice <- utils::menu(choices = vars_opt,
                     title = t,
                     graphics = FALSE)
      vars <- vars_list[[choice]]
    } else {

      vars <- vars_list[[1]]
    }
  }

  tab <- EMA_list_to_tibble(tab, vars)

  vars[pmatch("resp_time", vars)] <- "timestamp"
  for (v in vars) {
    if (all(strings_are_numeric(tab[[v]])))
      tab[[v]] <- as.numeric(tab[[v]])
  }

  ds <- attributes(tab)$dropped_students
  ds <- paste0(ds, collapse = ", ")

  if (ds > 0) {
    message(paste0("The students dropped ",
                   " with the choice of vars were numbers ", ds, "."))
  }

  class(tab) <- c("SL_tbl", class(tab))

  if ("timestamp" %in% names(tab)) {
    class(tab) <- c("EMA_SL_tbl", "timestamp_SL_tbl", class(tab))
  }

  return(tab)
}

EMA_to_list <- function(location, path) {

  `%>%` <- dplyr::`%>%`

  name <- get_name_from_path(path)
  if (name == "QR_Code") name <- "QR"
  paths <- generate_paths(location, path, name, ext = ".json")

  tab <- list()
  missing_tab <- 0
  for (i in 1:60) {
    if( file.exists(paths[i])
        && readLines(paths[i], 1, warn = FALSE) != "[]") {
      this_stud <- jsonlite::fromJSON(paths[i])
      this_stud$uid <- i - 1
      tab[[length(tab)+1]] <- this_stud
    } else {
      missing_tab <<- missing_tab + 1
    }
  }

  vars_present <- unique(lapply(tab, function(x){sort(names(x))} ))

  if ( length(vars_present) == 0 ) {
    stop(paste0("There was an error finding data. ",
                "Perhaps Check the location to ensure ",
                "it points to the top level of the StudentLife ",
                "dataset directory"))
  }

  for (i in 1:length(vars_present)) {
    if ( is.null(vars_present[[i]]) ) {
      vars_present[[i]] <- NULL; break
    }
  }

  ## Get EMA definition information
  if (name == "PAM") {

    EMA_questions <-
      paste0("Refer to: Pollak, J. P., Adams, P., & Gay, G.",
             " (2011, May). PAM: a photographic affect meter",
             " for frequent, in situ measurement of affect.",
             " In Proceedings of the SIGCHI conference on Human",
             " factors in computing systems (pp. 725-734). ACM.")

  } else if (name == "QR") {

    name <- "QR_Code"
    EMA_questions <- paste0(
                     "Classroom seating positions. See the layout at ",
                     "https://studentlife.cs.dartmouth.edu/dataset.html")

  } else {
    EMA_definition <- jsonlite::fromJSON(
      paste0(location,"/EMA/EMA_definition.json"), flatten = TRUE)
    EMA_names <- gsub("\\?", "_", EMA_definition$name)
    EMA_questions <- EMA_definition[
      pmatch(name, EMA_names),2][[1]]
    EMA_questions <- tibble::as_tibble(EMA_questions)
    EMA_questions <- EMA_questions[
      -pmatch("location", EMA_questions$question_id),]
  }

  tab <- structure(
    tab,
    vars_present = vars_present,
    EMA_name = name,
    EMA_questions = EMA_questions)

  return(tab)
}

EMA_list_to_tibble <- function(tab, vars = "resp_time") {

  tab_list <- tab

  # Drop lists not sharing variable names specified by parameter 'vars'
  null_ind <- c()
  for (i in 1:length(tab)) {
    if ( !all(vars %in% names(tab[[i]]) )) {
      null_ind <- c(null_ind, i)
    }
  }
  tab[null_ind] <- NULL

  # Drop all columns other than those specified by parameter 'vars'
  tab <- lapply(tab, function(x){
    dplyr::select(as.data.frame(x), vars)
  })

  `%>%` <- dplyr::`%>%`

  # Bind and make tibble
  tab <- tab %>%
    dplyr::bind_rows() %>%
    tibble::as_tibble()

  if ( "resp_time" %in% names(tab) )
    names(tab)[pmatch("resp_time", names(tab))] <- "timestamp"

  attr(tab_list, "dropped_students") <- null_ind - 1
  transfer_EMA_attrs(tab) <- tab_list

  return(tab)
}
